planner/core/rule_column_pruning.go

```diff
@@ -139,6 +139,11 @@ func (ls *LogicalSort) PruneColumns(parentUsedCols []*expression.Column) error {
 	for i := len(ls.ByItems) - 1; i >= 0; i-- {
 		cols := expression.ExtractColumns(ls.ByItems[i].Expr)
 		if len(cols) == 0 {
+			if sf, ok := ls.ByItems[i].Expr.(*expression.ScalarFunction); ok {
```

Comment:
SQL like `select ... from ... order by rand() + 1` also return data by random order.

Reply:
And there are more completed cases like `abs(rand())`, `floor(rand() * 100) %20`, `sysdate()`, which have the same problem.
There is a PR #10004 who introduce the concept `ConstItem()` for `expression`.
For each `expression`, it's better to check whether it is `ConstItem()`, and if it is not, we can't prune it.